cpp/src/hash/hashing.cu

```diff
@@ -0,0 +1,225 @@
+struct bitwise_partitioner
+{
+  bitwise_partitioner(size_type num_partitions) : divisor{(num_partitions - 1)} {
+    assert(is_power_two(num_partitions));
+  }
+
+  __host__ __device__
+  size_type operator()(hash_value_t hash_value) const {
+    return hash_value & (divisor);
+  }
+
+  const size_type divisor;
+};
+
+template <bool has_nulls>
+std::vector<std::unique_ptr<experimental::table>>
+hash_partition_table(table_view const& input,
+                     table_view const &table_to_hash,
+                     const size_type num_partitions,
+                     rmm::mr::device_memory_resource* mr,
+                     cudaStream_t stream)
+{
+  auto const num_rows = table_to_hash.num_rows();
+  auto row_partition_numbers = rmm::device_vector<size_type>(num_rows);
+
+  // Make an iterator to compute the hash over each row
+  auto const device_input = table_device_view::create(input, stream);
+  auto const hasher = experimental::row_hasher<MurmurHash3_32, has_nulls>(*device_input);
+  auto const hash_iterator = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(0), hasher);
+
+  // If the number of partitions is a power of two, we can compute the partition 
+  // number of each row more efficiently with bitwise operations
+  if (is_power_two(num_partitions)) {
+    // Compute which partition each row belongs to using bitwise partitioner
+    auto partitioner = bitwise_partitioner<hash_value_type>{num_partitions};
+    thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
+      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
+
+  } else {
+    // Compute which partition each row belongs to using modulo partitioner
+    auto partitioner = modulo_partitioner<hash_value_type>{num_partitions};
+    thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
+      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
+  }
+
+  // Scatter input rows to output partitions given the partition map
+  auto const partition_map = column_view(data_type(INT32), num_rows,
+    row_partition_numbers.data().get());
+  auto output = experimental::detail::scatter_to_tables(input, partition_map, mr, stream);
+
+  // Pad with empty tables if we have less than num_partitions
+  if (output.size() < static_cast<size_t>(num_partitions)) {
+    auto const old_size = output.size();
+    output.resize(num_partitions);
+    std::generate(output.begin() + old_size, output.end(),
+      [&input]() { return experimental::empty_like(input); });
+  }
+
+  return output;
+}
+
+}  // namespace
+
+namespace detail {
+
+std::vector<std::unique_ptr<experimental::table>>
+hash_partition(table_view const& input,
+               std::vector<size_type> const& columns_to_hash,
+               int num_partitions,
+               rmm::mr::device_memory_resource* mr,
+               cudaStream_t stream)
+{
+  CUDF_EXPECTS(columns_to_hash.size() > 0, "Need at least one column to hash");
```

Comment:
Typically in libcudf empty inputs result in empty outputs, not errors. This is important, for example, for clients like Dask (and I think Spark).

Reply:
Done